### ISSUE TITLE
Regression: path-based routing doesn't work as expected in 0.4.1 

### ISSUE THREAD

**Issue by james-xiang** [heptio/contour#338]
following the get started guid to make a deployment on minikube with NodePort.  The default backend configuration works fine. however, the customized path-based routing seems not working.
Following my deployment yaml files:
### Contour deployment which is based on contour-example-onrbac.yaml from user guide:
```
# This file is generated from the individual yaml files by deployment/render.sh.
# Do not edit this file directly but instead edit the source files and
# re-render.
apiVersion: v1
kind: Namespace
metadata:
  name: heptio-contour
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: contour
  namespace: heptio-contour
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  labels:
    app: contour
  name: contour
  namespace: heptio-contour
spec:
  selector:
    matchLabels:
      app: contour
  replicas: 1
  template:
    metadata:
      labels:
        app: contour
      annotations:
        prometheus.io/scrape: "true"
        prometheus.io/port: "9001"
        prometheus.io/path: "/stats"
        prometheus.io/format: "prometheus"
    spec:
      containers:
      - image: docker.io/envoyproxy/envoy-alpine:v1.6.0
        name: envoy
        ports:
        - containerPort: 8080
          name: http
        - containerPort: 8443
          name: https
        command: ["envoy"]
        args: ["-c", "/config/contour.yaml", "--service-cluster", "cluster0", "--service-node", "node0", "-l", "info", "--v2-config-only"]
        volumeMounts:
        - name: contour-config
          mountPath: /config
      - image: gcr.io/heptio-images/contour:v0.4.1
        imagePullPolicy: Always
        name: contour
        command: ["contour"]
        args: ["serve", "--incluster"]
      initContainers:
      - image: gcr.io/heptio-images/contour:v0.4.1
        imagePullPolicy: Always
        name: envoy-initconfig
        command: ["contour"]
        args: ["bootstrap", "/config/contour.yaml"]
        volumeMounts:
        - name: contour-config
          mountPath: /config
      volumes:
      - name: contour-config
        emptyDir: {}
      dnsPolicy: ClusterFirst
      serviceAccountName: contour
      terminationGracePeriodSeconds: 30
      # The affinity stanza below tells Kubernetes to try hard not to place 2 of
      # these pods on the same node.
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 100
            podAffinityTerm:
              labelSelector:
                matchLabels:
                  app: contour
              topologyKey: kubernetes.io/hostname
---
apiVersion: v1
kind: Service
metadata:
 name: contour
 namespace: heptio-contour
 # annotations:
  # This annotation puts the AWS ELB into "TCP" mode so that it does not
  # do HTTP negotiation for HTTPS connections at the ELB edge.
  # The downside of this is the remote IP address of all connections will
  # appear to be the internal address of the ELB. See docs/proxy-proto.md
  # for information about enabling the PROXY protocol on the ELB to recover
  # the original remote IP address.
  # service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
spec:
 ports:
 - port: 80
   name: http
   protocol: TCP
   targetPort: 8080
 - port: 443
   name: https
   protocol: TCP
   targetPort: 8443
 selector:
   app: contour
 type: NodePort
---
```
### example app deployment:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  labels:
    app: kuard
  name: kuard
spec:
  replicas: 3
  selector:
    matchLabels:
      app: kuard
  template:
    metadata:
      labels:
        app: kuard
    spec:
      containers:
      - image: gcr.io/kuar-demo/kuard-amd64:1
        name: kuard
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kuard
  name: kuard
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 8080
  selector:
    app: kuard
  sessionAffinity: None
  type: ClusterIP

```
### ingress deployment
```
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: kuard
  labels:
    app: kuard
spec:
  rules:
  - host: aptgw.test.com
    http:
      paths:
      - path: /kuard
        backend:
          serviceName: kuard
          servicePort: 80
```
### kubectl describe ingress kuard
```
Name:             kuard
Namespace:        default
Address:
Default backend:  default-http-backend:80 (<none>)
Rules:
  Host            Path  Backends
  ----            ----  --------
  aptgw.test.com
                  /kuard   kuard:80 (<none>)
Annotations:
Events:  <none>
```
### Envoy routes info from Envoy admin page
```
[
{
"version_info": "3",
"route_config_name": "ingress_http",
"config_source": {
 "api_config_source": {
  "api_type": "GRPC",
  "cluster_names": [
   "contour"
  ],
  "grpc_services": [
   {
    "envoy_grpc": {
     "cluster_name": "contour"
    }
   }
  ]
 }
}
,
"route_table_dump": {
 "name": "ingress_http",
 "virtual_hosts": [
  {
   "name": "aptgw.test.com",
   "domains": [
    "aptgw.test.com"
   ],
   "routes": [
    {
     "match": {
      "prefix": "/kuard"
     },
     "route": {
      "cluster": "default/kuard/80"
     }
    }
   ]
  }
 ]
}

}
]
```
It shows that Envoy doesn't get the correct cluster/backend server info.  It should be kuard:80 rather than default/kuard/80

**Comment by davecheney** [comment]
Thanks for raising this issue. Can you please clarify for me

- [ ] does your service work?
- [ ] if so, is this a bug report about the wrong details showing up in `kubectl describe`

**Comment by james-xiang** [comment]
@davecheney  http://aptgw.test.com/kuard doesn't work as well.

The example config without specify host name and path prefix works fine
spec:
  backend:
    serviceName: kuard
    servicePort: 80

 http://aptgw.test.com/ works fine

**Comment by davecheney** [comment]
Can you please follow the [troubleshooting document][0] and capture the contents of `curl http://127.0.0.1:9001/routes`. That will show the route set that has been applied.

[0]: https://github.com/heptio/contour/blob/master/docs/troubleshooting.md#access-the-envoy-admin-interface-remotely

**Comment by davecheney** [comment]
@james-xiang its a bit hard to follow what changes you have made. Can you please confirm for me

1. When you deploy this ingress
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: kuard
  labels:
    app: kuard
spec:
  backend:
    serviceName: kuard
    servicePort: 80
```
Then your service appears at http://aptgw.test.com/

However, when you deploy
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: kuard
  labels:
    app: kuard
spec:
  rules:
  - host: aptgw.test.com
    http:
      paths:
      - path: /kuard
        backend:
          serviceName: kuard
          servicePort: 80
```
Then http://aptgw.test.com/kuard does not work.

If this is correct, can you please paste the routes output as requested above as well as 
```
kubectl get ing kuard -o yaml
```
Thank you.

**Comment by james-xiang** [comment]
@davecheney Thanks for your reply.  following are  outputs:
### curl http://127.0.0.1:9001/routes
```
[
{
"version_info": "44",
"route_config_name": "ingress_http",
"config_source": {
 "api_config_source": {
  "api_type": "GRPC",
  "cluster_names": [
   "contour"
  ],
  "grpc_services": [
   {
    "envoy_grpc": {
     "cluster_name": "contour"
    }
   }
  ]
 }
}
,
"route_table_dump": {
 "name": "ingress_http",
 "virtual_hosts": [
  {
   "name": "apigw.test.com",
   "domains": [
    "apigw.test.com"
   ],
   "routes": [
    {
     "match": {
      "prefix": "/kuard"
     },
     "route": {
      "cluster": "default/kuard/80"
     }
    }
   ]
  }
 ]
}

}
]
```
### kubectl get ing kuard -o yaml
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"extensions/v1beta1","kind":"Ingress","metadata":{"annotations":{},"labels":{"app":"kuard"},"name":"kuard","namespace":"default"},"spec":{"rules":[{"host":"apigw.test.com","http":{"paths":[{"backend":{"serviceName":"kuard","servicePort":80},"path":"/kuard"}]}}]}}
  creationTimestamp: 2018-04-24T18:17:57Z
  generation: 1
  labels:
    app: kuard
  name: kuard
  namespace: default
  resourceVersion: "433023"
  selfLink: /apis/extensions/v1beta1/namespaces/default/ingresses/kuard
  uid: d0b3257f-47eb-11e8-85aa-08002768075a
spec:
  rules:
  - host: apigw.test.com
    http:
      paths:
      - backend:
          serviceName: kuard
          servicePort: 80
        path: /kuard
status:
  loadBalancer: {}
```
I have to mention that this host is actually the minikube node. In my laptop hosts file, I have a entry for it:
### cat /etc/hosts
```

##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost
255.255.255.255	broadcasthost
::1             localhost

# for testing
# minikube
192.168.99.100    apigw.test.com

```
### contour service url:
```
minikube service -n heptio-contour contour --url
http://192.168.99.100:30889
http://192.168.99.100:30693
```
I am using this url to access the kuard app via ingress controller
```
curl -i -v http://apigw.test.com:30889/kuard
*   Trying 192.168.99.100...
* TCP_NODELAY set
* Connected to apigw.test.com (192.168.99.100) port 30889 (#0)
> GET /kuard HTTP/1.1
> Host: apigw.test.com:30889
> User-Agent: curl/7.59.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
HTTP/1.1 404 Not Found
< date: Tue, 24 Apr 2018 18:39:52 GMT
date: Tue, 24 Apr 2018 18:39:52 GMT
< server: envoy
server: envoy
< content-length: 0
content-length: 0

<
* Connection #0 to host apigw.test.com left intact
```

**Comment by davecheney** [comment]
minikube, thats the missing piece of the puzzle. Because curl appends the port number to the url, it won't match the configuration in envoy. 

You'll have to use the `--header 'Host: apigw.test.com'` flag to correct the url generated by curl, `/etc/hosts` hacking won't help.

**Comment by davecheney** [comment]
Duplicate of #210
